File: src/leadforge/scrapers/yelp.py
import structlog
from leadforge.scrapers.base import BaseAPIClient
from leadforge.config import settings

logger = structlog.get_logger()

class YelpClient(BaseAPIClient):
    """Yelp Fusion API client."""

    def __init__(self):
        super().__init__(base_url="https://api.yelp.com/v3")

    async def search_business(self, name: str, location: str) -> dict | None:
        """Search for a business by name and location."""
        api_key = getattr(settings, 'YELP_API_KEY', '')
        if not api_key:
            logger.warning("yelp_api_key_not_set")
            return None

        try:
            client = await self._get_client()
            response = await client.get(
                "/businesses/search",
                params={"term": name, "location": location, "limit": 1},
                headers={"Authorization": f"Bearer {api_key}"},
            )
            response.raise_for_status()
            data = response.json()

            businesses = data.get("businesses", [])
            if not businesses:
                return None

            biz = businesses[0]
            return {
                "yelp_id": biz.get("id"),
                "yelp_review_count": biz.get("review_count", 0),
                "yelp_rating": biz.get("rating"),
                "yelp_price": biz.get("price"),
                "yelp_categories": [c.get("title") for c in biz.get("categories", [])],
            }
        except Exception as e:
            logger.warning("yelp_search_failed", error=str(e), name=name)
            return None

File: api/src/scrapers/yelp.ts
// =py scrapers/yelp
import { encodeParams, fetchJson } from './base';

export interface YelpEnv {
  // Python reads this with getattr(settings, "YELP_API_KEY", "") and Settings has no such field,
  // so the Python client is permanently disabled. Setting the Workers secret turns it on.
  YELP_API_KEY?: string;
}

export interface YelpBusiness {
  yelp_id: string | null;
  yelp_review_count: number;
  yelp_rating: number | null;
  yelp_price: string | null;
  yelp_categories: string[];
}

interface YelpApiBusiness {
  id?: string;
  review_count?: number;
  rating?: number;
  price?: string;
  categories?: { title?: string }[];
}

// =py search_business
export async function searchBusiness(env: YelpEnv, name: string, location: string): Promise<YelpBusiness | null> {
  if (!env.YELP_API_KEY) {
    console.warn('yelp_api_key_not_set');
    return null;
  }

  try {
    const params = encodeParams({ term: name, location, limit: '1' });
    const data = await fetchJson<{ businesses?: YelpApiBusiness[] }>(
      `https://api.yelp.com/v3/businesses/search?${params}`,
      { headers: { Authorization: `Bearer ${env.YELP_API_KEY}` } },
    );

    const business = data.businesses?.[0];
    if (!business) return null;

    return {
      yelp_id: business.id ?? null,
      yelp_review_count: business.review_count ?? 0,
      yelp_rating: business.rating ?? null,
      yelp_price: business.price ?? null,
      yelp_categories: (business.categories ?? []).map((c) => c.title).filter((t): t is string => Boolean(t)),
    };
  } catch (error) {
    console.warn('yelp_search_failed', { name, error: error instanceof Error ? error.message : String(error) });
    return null;
  }
}
